### PR TITLE
[styled-components] Allow static properties to be passed through

### LINKED ISSUE #32048: Extending styled-components interface with static properties
I am trying to add properties to the constructor produced by styled components using TypeScript.

What I want to do:

```tsx
type JSXButtonProps = JSX.IntrinsicElements['button'];

export interface ButtonProps extends JSXButtonProps {
  theme?: ThemeBase;
}

const ButtonInner = ({ theme, ...rest }: ButtonProps) => <button {...rest} />;

export const Button = styled(ButtonInner)(
  (props: ButtonProps): CSSEntry => get(props.theme, 'button.extend'),
);

Button.Primary = props => <Button />;
```

Which produces the error:

```
error TS2339: Property 'Primary' does not exist on type 'StyledComponent<FunctionComponent<ButtonProps>, ThemeBase, {}, never>'.
```

What works (casting to any):

```tsx
export const Button = styled(ButtonInner)(
  (props: ButtonProps): CSSEntry => get(props.theme, 'button.extend'),
);

(Button as any).Primary = props => <Button />;
```

I'd like to avoid this, so I looked into re-creating my own `ButtonInterface`, but the typings are so complex, I can't figure out where to start. For instance I see `StyledComponent<...>` but this is a generic type which cannot be extended, so that won't work. I've tried making the type and joining in the additional static class properties, but that hasn't worked either.

My latest attempt has been to do:

```tsx
interface ButtonAliases extends StyledComponentBase<any, any, any, any> {
  Primary: any;
  Secondary: any;
  Tertiary: any;
  Icon: any;
}

export const Button: ButtonAliases = styled(ButtonInner)(...);
```

But this does not work either, I get the error:

```
error TS2322: Type 'StyledComponent<({  theme, ...rest }: ButtonProps) => Element, ThemeBase, {}, never>' is not assignable to type 'ButtonAliases'.
```

Any help or suggestions would be great, thank you!

### PR DESCRIPTION
**Pull request by Lazyuki**
Allows styled components to retain static fields of the original component, which fixes #32048
Right now 
```jsx
class A extends React.Component {}
class B extends React.Component {
    static A = A;
}
const StyledB = styled(B)``;
<StyledB.A />
```
raises an error that says `Property 'A' does not exist on type 'StyledComponent<typeof B, DefaultTheme, {}, never>'` but if this was plain JS it works as intended. 

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: #32048

### CODE REVIEW

**Review by eps1lon — COMMENTED** [review]
I guess this is fine. Though it seems like react statics like `propTypes` and `getDerivedStateFromProps` are now hoisted as well to the resulting type.

Would you be so kind and link the official documentation that describes hoisting of statics?

**Comment by Lazyuki** [comment]
I suppose since `styled-components` already uses `hoist-non-react-statics` I should also do that here.

**Comment by eps1lon** [comment]
> I suppose since `styled-components` already uses `hoist-non-react-statics` I should also do that here.

It was more about referencing that feature. I never used statics on styled-components so I'm not aware that they even do that.

File: types/styled-components/index.d.ts [hunk 1]
```diff
@@ -21,6 +22,7 @@ declare global {
 
 import * as CSS from "csstype";
 import * as React from "react";
+import hoistNonReactStatics = require('hoist-non-react-statics');
```
Comment (eps1lon) [inline types/styled-components/index.d.ts]: Prefer `import ... from` to be consistent with previous imports.

**Review by eps1lon — CHANGES_REQUESTED** [review]

File: types/styled-components/index.d.ts (hunk 1 above)
Comment (Jessidhia) [reply types/styled-components/index.d.ts]: Actually I'd prefer to change the other imports to `import = require`. `react` is not an ES module, and as long as `esModuleInterop` doesn't default to `true` for `.d.ts` parsing it's better to be explicit about the commonjs import.

Comment (eps1lon) [reply types/styled-components/index.d.ts]: I'd agree but that is a) part of another change and b) inconsistent with virtually everybody else. That boat has long sailed where we can insist to not import react as an esmodule. `import * as React` is close enough. `import React = require()` is very uncommon syntax.

**Review by eps1lon — APPROVED** [review]

**Comment by eps1lon** [comment]
@Lazyuki `npm run test styled-components` should reveal potential issues. Just a tiny syntax error.

**Comment by Lazyuki** [comment]
@eps1lon Yeah I tried that but it errors out every time with `Error: .DS_Store: filesystem doesn't support paths of the form './x'.` and I can't seem to find any reference to this issue online. I tried deleting `.DS_Store` and also modifying the test script but didn't work...

**Comment by Lazyuki** [comment]
Could I get the `Revision needed` label removed? Seems like it hasn't been removed automatically after the revision was done.

**Comment by orta** [comment]
Looks good 👍

**Pull request merged**